fs/expose/dokan/__init__.py

```diff
@@ -359,111 +404,104 @@ def _check_lock(self, path, offset, length, info, locks=None):
 
         This method implements basic lock checking.  It checks all the locks
         held against the given file, and if any overlap the given byte range
-        then it returns -ERROR_LOCKED.  If the range is not locked, it will
+        then it returns STATUS_LOCK_NOT_GRANTED.  If the range is not locked, it will
         return zero.
         """
         if locks is None:
             with self._files_lock:
                 try:
                     locks = self._active_locks[path]
                 except KeyError:
-                    return 0
+                    return STATUS_SUCCESS
         for (lh, lstart, lend) in locks:
             if info is not None and info.contents.Context == lh:
                 continue
             if lstart >= offset + length:
                 continue
             if lend <= offset:
                 continue
-            return -ERROR_LOCKED
-        return 0
+            return STATUS_LOCK_NOT_GRANTED
+        return STATUS_SUCCESS
 
     @timeout_protect
     @handle_fs_errors
-    def CreateFile(self, path, access, sharing, disposition, flags, info):
-        path = normpath(path)
+    def ZwCreateFile(self, path, securitycontext, access, attribute, sharing, disposition, options, info):
+        path = self._dokanpath2pyfs(path)
         #  Can't open files that are pending delete.
         if self._is_pending_delete(path):
-            return -ERROR_ACCESS_DENIED
-        # If no access rights are requestsed, only basic metadata is queried.
-        if not access:
-            if self.fs.isdir(path):
-                info.contents.IsDirectory = True
-            elif not self.fs.exists(path):
-                raise ResourceNotFoundError(path)
-            return
-        #  This is where we'd convert the access mask into an appropriate
-        #  mode string.  Unfortunately, I can't seem to work out all the
-        #  details.  I swear MS Word is trying to write to files that it
-        #  opens without asking for write permission.
-        #  For now, just set the mode based on disposition flag.
-        retcode = 0
-        if disposition == CREATE_ALWAYS:
-            if self.fs.exists(path):
-                retcode = ERROR_ALREADY_EXISTS
-            mode = "w+b"
-        elif disposition == OPEN_ALWAYS:
-            if not self.fs.exists(path):
+            return STATUS_ACCESS_DENIED
+
+        retcode = STATUS_SUCCESS
+        if info.contents.IsDirectory:
```

Comment:
While toying with Dokan, I had to change it like this:
```python
if self.fs.isdir(path) or info.contents.IsDirectory:
    info.contents.IsDirectory = True
```
Otherwise Windows would mistake directories to files and vice versa:
```python
>>> os.path.isfile(r'P:\\')
True
>>> os.path.isdir(r'P:\\')
False
```